File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        ans = ListNode(-1)
        head = ans
        temp1 = l1;temp2 = l2
        prevcarry=0

        while (temp1 or temp2):
            if (temp1):
                d1 = temp1.val
                temp1=temp1.next
            else: d1=0
            if (temp2):
                d2 = temp2.val
                temp2=temp2.next
            else: d2=0

            s = d1+d2+prevcarry
            curcarry = s//10
            s = s%10

            ans.next = ListNode(s)
            ans = ans.next

            prevcarry = curcarry

        if curcarry:
            ans.next = ListNode(curcarry)
            
        head = head.next
        return head
        
            
                
            
            

        
        
        
#   1 1 1 1 1 1       
#   9 9 9 9 9 9 9
#   9 9 9 9 9 9 9
#   1 9 9 9 9 9 9 8   








        
        

        

File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {

public:
    int Addition(int val1, int val2,int carry){
            int s = (val1+val2)+carry;
            return s;
        }
   
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head1 = l1;
        ListNode* head2 = l2;
        ListNode* prev;
        int carry=0;
        int s;
        while(l2!=NULL && l1!=NULL){
            s = Addition(l1->val,l2->val,carry);
            carry = (s)/10;
            s = s%10;
            prev = l1;
            l1->val = s;
            l1 = l1->next;
            l2 = l2->next;
        }
        
        if (l1==NULL && l2==NULL){
            if (carry){
                l1=prev;
                l1->next = new ListNode(carry);
            }
        }
        else if (l1!=NULL){
            while(l1!=NULL){
                s = Addition(l1->val,0,carry);
                carry = (s)/10;
                s = s%10;
                prev = l1;
                l1->val = s;
                l1 = l1->next;
            }
            if (carry){
                l1=prev;
                l1->next = new ListNode(carry);
            }
        }
        else if (l2!=NULL){
            l1 = prev;
            while(l2!=NULL){
                s = Addition(0,l2->val,carry);
                carry = (s)/10;
                s = s%10;
                l1->next = new ListNode(s);
                l1 = l1->next;
                l2 = l2->next;
            }
            if (carry){
                l1->next = new ListNode(carry);
            }
        }
        return head1;
    }
};